feat(app-delivery) IAM policy for deploy stack

moofish32 opened the pull request:
* The changeset and apply changeset may need specific IAM permissions
 and the user can now customize them via `deployStackAction.role`
 * Document updates for proper build stage configuration
 * Fixes #1151


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

eladb commented:
@skinny85  can you take a look please?

rix0rrr commented on packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts:
```diff
@@ -41,6 +41,29 @@ export interface PipelineDeployStackActionProps {
+  /**
+   * The CloudFormation Capabilities enabled for the ChangeSet.
+   *
+   * @default None
```
I guess this is a request for upstream change then.

@skinny85  ?

RomainMuller replied:
The CloudFormation documentation states that `NamedIAM` implies `IAM`, and that you need not specify both.

rix0rrr replied:
That future-proofing is exactly what I'm opposing to.

skinny85 replied:
@moofish32, I believe @rix0rrr means "anonymous IAM permissions" (as opposed to "named IAM permissions").

I agree wholeheartedly with Rico and Romain here. In general, we don't do these kind of future proofing in the CDK - we try to give a nicer API for the common case.

So, I would proceed as follows here:

1. Introduce a new `enum` in this module, with 3 values: `None`, `AnonymousIAM`, `NamedIAM`.
2. Make `capabilities` in `PipelineDeployStackActionProps` be of this new `enum` type (*not* an array), with default `AnonymousIAM`.
3. After this PR, open a separate PR to move this enum to the `aws-cloudformation` package, and change the API of the CFN Actions themselves to use it.

Does this make sense?

moofish32 replied:
If I can do it all in this PR should I? I haven't looked at how difficult this would be plumb.

skinny85 replied:
Probably easier to leave it out of this PR, as it's a breaking change. But knock yourself out if you want - I don't think it will be super difficult :).

RomainMuller commented on packages/@aws-cdk/app-delivery/README.md:
````diff
@@ -69,26 +73,41 @@ const deployStage = pipeline.addStage('Deploy');
 ```
 
 #### `buildspec.yml`
+The repository should contain a file at the root level named `buildspec.yml`.
````
That's only true if you don't specify an inline buildspec with your CodeBuild project, so I wouldn't phrase it like so.

RomainMuller commented on packages/@aws-cdk/app-delivery/README.md:
```diff
@@ -69,26 +73,41 @@ const deployStage = pipeline.addStage('Deploy');
 For example, a *TypeScript* or *Javascript* CDK App can add the following `buildspec.yml` at the root of the repository
 configured in the `Source` stage:
 
+Example contents of `buildspec.yml` (note: `buildspec.yaml` is not compatible)
```
The precision on `.yaml` is definitely a good one - I'd have moved it in the paragraph just above.

RomainMuller commented on packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts:
```diff
@@ -41,6 +41,29 @@ export interface PipelineDeployStackActionProps {
   executeChangeSetRunOrder?: number;
+
+  /**
+   * The role to use when creating and executing the ChangeSet.
```
I would copy the comment from the `PipelineCreateReplaceChangeSetActionProps` attribute (same for the other two). I think they're pretty detailed and they avoid insisting so much on the implementation detail here (customer is really just interesting at deploying a stack - the fact there is a ChangeSet being created, then executed, is mostly irrelevant).

RomainMuller commented on packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts:
```diff
@@ -69,16 +98,23 @@ export class PipelineDeployStackAction extends cdk.Construct {
-    new cfn.PipelineCreateReplaceChangeSetAction(this, 'ChangeSet', {
+    this.stack = props.stack;
+
+    const fullPermissions = props.fullPermissions === true;
```
I don't think this is needed. `fullPermissions` defaults to `false` in `PipelineCreateReplaceChangeSetAction` (and I reckon we're quite unlikely to change this default, as it has pretty drastic security implications).

RomainMuller commented on packages/@aws-cdk/app-delivery/package.json:
```diff
@@ -63,6 +65,8 @@
   ],
   "peerDependencies": {
     "@aws-cdk/aws-codepipeline-api": "^0.16.0",
+    "@aws-cdk/aws-cloudformation": "^0.16.0",
```
You'll want to make sure you merge in the `0.17.0` version bump & update those :)

RomainMuller commented on packages/@aws-cdk/app-delivery/README.md:
```diff
@@ -52,9 +52,13 @@ const source = new codepipeline.GitHubSourceAction(pipelineStack, 'GitHub', {
 const project = new codebuild.PipelineProject(pipelineStack, 'CodeBuild', {
-  /* ... */
+  environment: {
+    buildImage: codebuild.LinuxBuildImage.UBUNTU_14_04_NODEJS_10_1_0,
```
I wouldn't specify this in the `README.md`, it could be distracting (or looking like it needs to be that).

moofish32 replied:
Made a comment instead as an example.

skinny85 commented on packages/@aws-cdk/app-delivery/package.json:
```diff
@@ -46,6 +51,21 @@
+    "@aws-cdk/aws-codepipeline": "^0.17.0",
+    "@aws-cdk/aws-s3": "^0.17.0",
+    "cdk-build-tools": "^0.17.0",
+    "cdk-integ-tools": "^0.17.0",
```
Some dependencies are duplicated here (`cdk-build-tools`, `cdk-integ-tools`, `@aws-cdk/aws-s3`, `@aws-cdk/aws-codepipeline`) - I would remove the extra ones  so that the `@aws-cdk` namespace ones are grouped at the beginning.

skinny85 commented on packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts:
```diff
@@ -69,16 +117,21 @@ export class PipelineDeployStackAction extends cdk.Construct {
     const changeSetName = props.changeSetName || 'CDK-CodePipeline-ChangeSet';
 
-    new cfn.PipelineCreateReplaceChangeSetAction(this, 'ChangeSet', {
+    this.stack = props.stack;
```
Nit: I would not move this line (makes the diff look weird).

rix0rrr commented:
I will leave it to @RomainMuller and @skinny85 to merge this PR if they're satisfied.

RomainMuller commented:
I think we'll be good to merge once @skinny85's suggestion with respects to defining an `enum` locally is implemented.

moofish32 commented:
I have updated the commit (via squash/rebase) and included this breaking change. If we are going to break let's just get on with it.

```
* The changeset and apply changeset can now apply role IAM permissions,
 and CloudFormation Capabilities
 * Updated CloudFormationCapabilities enum to include `None` and renamed `IAM` to `AnonymousIAM`.
 * Document updates for proper build stage configuration
 * Fixes #1151

BREAKING CHANGE: `CloudFormationCapabilities.IAM` renamed to
`CloudFormation.AnonymousIAM` and `PipelineCloudFormationDeployActionProps.capabilities?: CloudFormationCapabilities[]` has been changed to
`PipelineCloudFormationDeployActionProps.capabilities?:
CloudFormationCapabilities` no longer an array.

```

moofish32 commented on packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts:
```diff
@@ -41,6 +41,50 @@ export interface PipelineDeployStackActionProps {
+   *
+   * @default false
+   */
+  fullPermissions?: boolean;
```
I went with `adminPermissions` because when I put in `administrator` I asked my self of what? I thought `adminPermissions` would be more clear, if you are set on `administrator` let me know.

moofish32 commented on packages/@aws-cdk/app-delivery/README.md:
````diff
@@ -69,26 +77,42 @@ const deployStage = pipeline.addStage('Deploy');
+ // ... addAction('actions that you need').
+ // add resource
+);
 ```
````
I didn't add another example but I added text to describe the option; does this work? Or do we want a separate example.

skinny85 replied:
I think, since both examples are the same, I would add a `fullPermissions` example as well.

skinny85 replied:
Yep, that's what I meant 👍🏻

skinny85 commented on packages/@aws-cdk/app-delivery/README.md:
```diff
@@ -69,26 +77,48 @@ const deployStage = pipeline.addStage('Deploy');
+// Add the necessary permissions for you service deploy action. This role is
+// is passed to CloudFormation and needs the permissions necessary to deploy
+// stack. Alternatively you can enable [Administrator](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_job-functions.html#jf_administrator
+) permissions above, 
```
Huh. This shows up as a new line without comments. Perhaps it's just GitHub's diff displaying it weirdly.

skinny85 commented on packages/@aws-cdk/app-delivery/README.md:
```diff
@@ -69,26 +77,48 @@ const deployStage = pipeline.addStage('Deploy');
 For example, a *TypeScript* or *Javascript* CDK App can add the following `buildspec.yml` at the root of the repository
 configured in the `Source` stage:
 
+Example contents of `buildspec.yml`.
```
Example contents of `buildspec.yml`:
// empty line here
```yml
version: 0.2
phases:
```
...

skinny85 replied:
I think this now clashes with the sentence above? "For example, a *TypeScript* or *Javascript* CDK App can add the following `buildspec.yml` at the root of the repository configured in the `Source` stage:". Perhaps it always did, I just never noticed ;p.

skinny85 commented on packages/@aws-cdk/app-delivery/lib/pipeline-deploy-stack-action.ts:
```diff
@@ -41,6 +41,47 @@ export interface PipelineDeployStackActionProps {
+   * Acknowledge certain changes made as part of deployment
+   *
+   * For stacks that contain certain resources, explicit acknowledgement that AWS CloudFormation
+   * might create or update those resources. For example, you must specify CAPABILITY_IAM if your
```
I believe `CAPABILITY_IAM` is not the correct constant name anymore?

skinny85 commented on packages/@aws-cdk/app-delivery/test/test.pipeline-deploy-stack-action.ts:
```diff
@@ -51,14 +62,170 @@ export = nodeunit.testCase({
+      addAction('ec2:DescribeSecurityGroups').
+      addAction('ec2:CreateSecurityGroup').
+      addAction('ec2:RevokeSecurityGroupEgress').
+      addAction('ec2:RevokeSecurityGroupIngress').
```
Feel free to use the `addActions` method here, that takes a varargs argument, will be more concise.

moofish32 replied:
OH... I didn't see there was one!

skinny85 commented on packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts:
```diff
@@ -494,3 +502,16 @@ function stackArnFromProps(props: { stackName: string, region?: string }): strin
+    // else capabilities are defined use them
+    return capabilities;
+  }
+}
```
Sorry, maybe I wasn't clear before.

The capabilities should default to `AnonymousIAM` only in the `app-delivery` construct.

In the CFN Actions, the default should be `undefined` (that is, none), like it is now.

moofish32 replied:
I wondered about this change. With `adminPermissions` being required it seems a bit odd this default is allow IAM, but it's a pretty heavy coin flip considering it's rare not to deploy a role with a service.

skinny85 commented on packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts:
```diff
@@ -151,10 +151,8 @@ export interface PipelineCloudFormationDeployActionProps extends PipelineCloudFo
-   * @default false
    */
-  fullPermissions?: boolean;
+  adminPermissions: boolean;
```
Why did we change this to required? IMO, it should still default to `false`.

skinny85 replied:
Ah, I see that was @eladb's comment.

Feel free to leave it required then.

moofish32 commented:
Thanks for the conflict resolution @RomainMuller. @skinny85 did I get all the fixes in here?

skinny85 commented on packages/@aws-cdk/app-delivery/README.md:
```diff
@@ -69,26 +77,44 @@ const deployStage = pipeline.addStage('Deploy');
+ // new iam.PolicyStatement().
+ // ... addAction('actions that you need').
+ // add resource
+);
```
I wouldn't comment out the whole thing, but instead just the usecase-specific part:

```ts
deployServiceAAction.addToRolePolicy(
  new iam.PolicyStatement()
    .addAction('service:SomeAction')
    .addResource(myResource.myResourceArn)
    // add more Action(s) and/or Resource(s) here, as needed
);
```

skinny85 commented on packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts:
```diff
@@ -118,24 +118,24 @@ export interface PipelineCloudFormationDeployActionProps extends PipelineCloudFo
+   * might create or update those resources. For example, you must specify `AnonymousIAM` or `NamedIAM`
+   * if your stack template contains AWS Identity and Access Management (IAM) resources. For more
+   * information
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities
```
I don't think this will render well (the sentence will be cut off, and the @see will be below the text paragraph. Feel free to use Markdown links here: `For more information, [see the AWS documentation](https://docs.aws.amazon.com/somepath)`.

skinny85 replied:
Ok, this way is fine 🙂

skinny85 commented on packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts:
```diff
@@ -366,8 +364,17 @@ export enum CloudFormationCapabilities {
+   * @link https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-iam-template.html#using-iam-capabilities
    */
-  NamedIAM = 'CAPABILITY_NAMED_IAM'
+  None = '',
```
I would actually make `None` the first enum value (I would sort them in the order of increasing capabilities).

skinny85 commented on packages/@aws-cdk/aws-codepipeline/test/integ.cfn-template-from-repo.lit.expected.json:
```diff
@@ -322,4 +322,4 @@
     }
   }
-}
+}
```
Can you try to revert this change? I think the integ tests will pass without this ordering change.

skinny85 commented on packages/@aws-cdk/aws-codepipeline/test/integ.lambda-pipeline.expected.json:
```diff
@@ -181,7 +175,13 @@
+            "Ref": "PipelineArtifactsBucket22248F97"
+          },
+          "Type": "S3"
+        }
```
Same comment here (try to revert this change and see if the compare tests still pass, they should).

skinny85 commented on packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-cfn.expected.json:
```diff
@@ -254,4 +254,4 @@
     }
   }
-}
+}
```
Same comment here as in the other 2 expected JSON files.

skinny85 commented on packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-build-multiple-inputs-outputs.expected.json:
```diff
@@ -355,7 +349,13 @@
+            "Ref": "MyBucketF68F3FF0"
+          },
+          "Type": "S3"
+        }
```
Same comment here as in the other 3 expected JSON files.

skinny85 commented on packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-commit-build.expected.json:
```diff
@@ -219,7 +213,13 @@
+            "Ref": "PipelineArtifactsBucket22248F97"
+          },
+          "Type": "S3"
+        }
```
Same comment here as in the other 4 expected JSON files.

skinny85 commented on packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-commit.expected.json:
```diff
@@ -221,7 +215,13 @@
+            "Ref": "PipelineArtifactsBucket22248F97"
+          },
+          "Type": "S3"
+        }
```
Same comment here as in the other 5 expected JSON files.

skinny85 commented on packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-code-deploy.expected.json:
```diff
@@ -328,7 +322,13 @@
+            "Ref": "CodeDeployPipelineIntegTest9F618D61"
+          },
+          "Type": "S3"
+        }
```
Same comment here as in the other 6 expected JSON files.

skinny85 commented on packages/@aws-cdk/aws-codepipeline/test/integ.pipeline-events.expected.json:
```diff
@@ -103,12 +103,6 @@
     "MyPipelineAED38ECF": {
       "Type": "AWS::CodePipeline::Pipeline",
       "Properties": {
-        "ArtifactStore": {
```
Same comment here as in the other 7 expected JSON files.

skinny85 commented on packages/@aws-cdk/aws-cloudformation/lib/pipeline-actions.ts:
```diff
@@ -198,12 +196,11 @@ export abstract class PipelineCloudFormationDeployAction extends PipelineCloudFo
-      // This must be a string, so flatten the list to a comma-separated string.
-      Capabilities: (capabilities && capabilities.join(',')) || undefined,
+      // None evaluates to empty string which is falsey and results in undefined
+      Capabilities: capabilities && capabilities.toString() || undefined,
```
Oooh, you're relying on the fact that `&&` binds stronger than `||` - that is pure evil :). Mind adding some parenthesis around the 'and' expression?

moofish32 commented:
alright @skinny85 one more pass when ever you are back from the holiday.
